### ISSUE TITLE
Unintended FastBoot dependencies under Embroider

### ISSUE THREAD

**Issue by bobisjan** [embroider-build/embroider#685]
Hi,

I have a question about content of `dist/package.js` under FastBoot. It looks that the `dist/package.json` is _extended_ version of `package.json` with FastBoot extras.

In classic build the `dist/package.json` contains `dependencies` based `fastbootDependencies` in `package.json`, then `npm install` is required to run in `dist/` (eq. [fastboot-s3-downloader](https://github.com/ember-fastboot/fastboot-s3-downloader/blob/master/index.js#L110-L114)).

```
{
  "dependencies": {
    "abortcontroller-polyfill": "^1.4.0",
    "node-fetch": "^2.6.1"
  }
}
```

When using Embroider the `dist/package.json` contains also references for in-repo add-ons, thus later run of `npm install` fails.

```
{
  "dependencies": {
    "aws": "*",
    "build": "*",
    "dom-attrs": "*",
    "gtm": "*",
    "headers": "*",
    "manifest": "*",
    "one-signal": "*",
    "redirect-support": "*",
    "sitemap": "*",
    "test-fastboot": "*",
    "version": "*",
    "abortcontroller-polyfill": "^1.4.0",
    "node-fetch": "^2.6.1"
  }
}
```

How these extra dependencies should be handled in FastBoot?

**Comment by ef4** [comment]
Under embroider, the fastboot build is already ready to run and doesn't need a separate npm install.

Any dependency that you import from fastboot-only code will be built into your app's fastboot-only bundles.

**Comment by bobisjan** [comment]
Yes, but that would require to move away from `FastBoot.require('3rd')` (including add-ons) before Embroider can be used in app?

**Comment by ef4** [comment]
True and we want apps to be able to switch back and forth without incompatible changes to the app code, so this is probably a good feature to add to embroider. I think all that's needed is transpiling `FastBoot.require('3rd')` to `importSync('3rd')`.

**Comment by bobisjan** [comment]
> True and we want apps to be able to switch back and forth without incompatible changes to the app code, so this is probably a good feature to add to embroider.

👍 I'm happy to take a look at this feature.

> I think all that's needed is transpiling FastBoot.require('3rd') to importSync('3rd').

Sounds good. What about Node.js built-ins like `fs`? Will be `importSync` able to handle these too?

**Comment by ef4** [comment]
Ah, that is a good question. Some additional config is going to be required to make that work. You would need to tell webpack to treat them as `externals`. Something like this:

```js
return require('@embroider/compat').compatBuild(app, Webpack, {
  packagerOptions: {
     webpackConfig: {
       externals: {
         fs: "require('fs')"
       }
     }
  }
});
```

**Comment by bobisjan** [comment]
Thanks for clarification, before I dive in I've made an example of the changes.

```javascript
// before
import Route from '@ember/routing/route';
import { inject as service } from '@ember/service';

export default IndexRoute extends Route {
  @service fastboot;

  model() {
    if (this.fastboot.isFastBoot) { // or typeof FastBoot !== 'undefined'
      let process = FastBoot.require('process'); // Node.js built-in
      let something = FastBoot.require('something'); // 3rd package
      
      return `${process.title}:${something}`;
    } else {
      return 'other';
    }
  }
}

// after
import Route from '@ember/routing/route';
import { inject as service } from '@ember/service';
import { importSync } from '@embroider/macros';

export default IndexRoute extends Route {
  @service fastboot;

  model() {
    if (this.fastboot.isFastBoot) { // or typeof FastBoot !== 'undefined'
      let process = require('process'); // Node.js built-in
      let something = importSync('something'); // 3rd package

      return `${process.title}:${something}`;
    } else {
      return 'other';
    }
  }
}
```


1.  https://github.com/embroider-build/embroider/blob/936367159b370492b0bff4b67373d1da29d49fd3/packages/core/src/babel-plugin-adjust-imports.ts is this correct place to the job?
2. `this.fastboot.isFastBoot` or `typeof FastBoot !== 'undefined'` will work properly or `macroCondition(getGlobalConfig().fastboot?.isRunning)` should be used instead to avoid bundling `something` into browser assets?
